include/grpcpp/impl/codegen/message_allocator.h

```diff
@@ -22,31 +22,44 @@
 namespace grpc {
 namespace experimental {
 
-// This is per rpc struct for the allocator. We can potentially put the grpc
-// call arena in here in the future.
+// NOTE: This is an API for advanced users who need custom allocators.
+// Per rpc struct for the allocator. This is the interface to return to user.
+class RpcAllocatorState {
+ public:
+  virtual ~RpcAllocatorState() = default;
+  // Optionally deallocate request early to reduce the size of working set.
+  // A custom MessageAllocator needs to be registered to make use of this.
+  // This is not abstract because implementing it is optional.
+  virtual void FreeRequest() {}
+};
+
+// This is the interface returned by the allocator.
+// grpc library will call the methods to get request/response pointers and to
+// release the object when it is done.
 template <typename RequestT, typename ResponseT>
-struct RpcAllocatorInfo {
-  RequestT* request;
-  ResponseT* response;
-  // per rpc allocator internal state. MessageAllocator can set it when
-  // AllocateMessages is called and use it later.
-  void* allocator_state;
+class MessageHolder : public RpcAllocatorState {
+ public:
+  virtual void Release() { delete this; }
+  RequestT* request() { return request_; }
+  ResponseT* response() { return response_; }
+  void set_request(RequestT* request) { request_ = request; }
+  void set_response(ResponseT* response) { response_ = response; }
+
```

Comment:
setters as protected since they are only going to be set by a derived class?

Reply:
Done.